File: cpp/organism/plant/species/SosnowskyHogweed.cpp
//
// Created by fgrel on 18.04.2025.
//

#include "SosnowskyHogweed.h"
#include "../../animal/Animal.h"

Organism* SosnowskyHogweed::clone() const {
  return new SosnowskyHogweed(*this); // Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

void SosnowskyHogweed::reproduce() {
  Point newPosition = getNewPosition();
  if (newPosition == position)
    return; // Ensure a new position is valid

  world.killPlantsOnPosition(newPosition);
  world.queueOrganismAddition(new SosnowskyHogweed(newPosition, world));
}

double SosnowskyHogweed::getSpreadProbability() const {
  return spreadProbability;
}

bool SosnowskyHogweed::collision(Organism &other) {
  world.getLogger().logEvent("SosnowskyHogweed has killed  " +
                              other.getSymbol() + ".");
  other.die();
  return false;
}

void SosnowskyHogweed::killNearbyAnimals() {
  std::vector<Point> neighbors = world.getNeighboringPositions(position);
  for (const Point &neighbor : neighbors) {
    std::vector<Organism *> organisms =
        world.getOrganismsAt(neighbor);
    for (const auto &organism : organisms) {
      if (organism && dynamic_cast<Animal *>(organism)) { // Sprawdź, czy to zwierzę
        world.getLogger().logEvent("Sosnowsky Hogweed zabił " + organism->getSymbol() + " na pozycji (" +
                                   std::to_string(neighbor.x) + ", " + std::to_string(neighbor.y) + ").");
        organism->die();
      }
    }
  }
}

File: java/src/organism/plant/species/SosnowskyHogweed.java
package organism.plant.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.animal.Animal;
import organism.plant.Plant;
import world.World;

import java.util.List;

public class SosnowskyHogweed extends Plant {

    public SosnowskyHogweed(int x, int y, World world) {
        super(x, y, Constants.Plant.SosnowskyHogweed.Symbol, Constants.Plant.SosnowskyHogweed.Strength, Constants.Plant.Initiative, world);
    }

    @Override
    public void action() {
        increaseAge();
        killNearbyAnimals();
        if (Math.random() <= getSpreadProbability()) {
            reproduce();
        }
    }

    @Override
    protected void reproduce() {
        int[] newPosition = getNewPosition();
        if (newPosition[0] == x && newPosition[1] == y) return;
        world.addOrganism(new SosnowskyHogweed(newPosition[0], newPosition[1], world));
        EventLogger.getInstance().log("Sosnowsky Hogweed spread to (" + newPosition[0] + ", " + newPosition[1] + ")");
    }

    @Override
    protected double getSpreadProbability() {
        return Constants.Plant.SosnowskyHogweed.SpreadProbability;
    }

    @Override
    public boolean collision(Organism other) {
        EventLogger.getInstance().log("Sosnowsky Hogweed has killed " + other.getClass().getName() + " at (" + x + ", " + y + ")");
        other.die();
        return false;
    }

    private void killNearbyAnimals() {
        List<int[]> neighbors = world.getNeighbors(x, y);
        for (int[] pos : neighbors) {
            Organism org = world.getOrganismAt(pos[0], pos[1]);
            if (org instanceof Animal) {
                EventLogger.getInstance().log("Sosnowsky Hogweed killed " + org.getClass().getName() + " at (" + pos[0] + ", " + pos[1] + ")");
                org.die();
            }
        }
    }
}